### PR TITLE
Nicer displayException for io-sim's Failure

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

File: io-sim/src/Control/Monad/IOSim.hs [hunk 1]
```diff
@@ -111,7 +112,14 @@ data Failure =
      | FailureSloppyShutdown [LabeledThread]
   deriving Show
 
-instance Exception Failure
+instance Exception Failure where
+    displayException (FailureException err) = displayException  err
+    displayException FailureDeadlock = "<<io-sim deadlock>>"
+    displayException (FailureSloppyShutdown threads) =
+      concat [ "<<sloppy shutdown: "
```
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim.hs:119]: Do we want to include "io-sim" in both exception messages (deadlock and sloppy shutdown), or neither?

**Review by dcoutts — APPROVED** [review]

File: io-sim/src/Control/Monad/IOSim.hs (hunk 1 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSim.hs:119]: I prefer both.

Comment (coot) [reply io-sim/src/Control/Monad/IOSim.hs]: And what about `Failure err`, maybe `"<<io-sim failure: " ++ displayException err ++ ">>"`?

Comment (coot) [reply io-sim/src/Control/Monad/IOSim.hs]: I decided against it, that would not work well if `displayException err` contained new line characters.

**Comment by coot** [comment]
bors merge

**Pull request merged**